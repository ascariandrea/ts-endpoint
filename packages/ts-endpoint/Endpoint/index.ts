import * as t from 'io-ts';

export const HTTP_METHODS = {
  OPTIONS: 'OPTIONS' as 'OPTIONS',
  HEAD: 'HEAD' as 'HEAD',
  GET: 'GET' as 'GET',
  POST: 'POST' as 'POST',
  PUT: 'PUT' as 'PUT',
  PATCH: 'PATCH' as 'PATCH',
  DELETE: 'DELETE' as 'DELETE',
};
export type HTTP_METHODS = typeof HTTP_METHODS;

export const HTTPMethod = t.keyof(
  {
    OPTIONS: null,
    HEAD: null,
    GET: null,
    POST: null,
    PUT: null,
    PATCH: null,
    DELETE: null,
  },
  'HTTPMethod'
);

export type HTTPMethod = t.TypeOf<typeof HTTPMethod>;
/**
 * Endpoint options type
 */
export interface EndpointOptions {
  stringifyBody?: boolean;
}

/**
 * Represents an endpoint of our API
 *
 * @typeparam O Endpoint options
 * @typeparam P Endpoint `Params` type
 * @typeparam H Endpoint `Headers` type
 * @typeparam Q Endpoint `Query` type
 * @typeparam B Endpoint `Body` type
 * @typeparam R Endpoint `Output` type
 */
export interface Endpoint<O extends EndpointOptions, P, H, Q, B, R> {
  Method: HTTPMethod;
  // `Path` is a function that accept `Input.Params` when is defined or no argument when is not defined
  Path: P extends Record<any, any> ? (args: Record<keyof P, string>) => string : () => string;
  Opts: O;
  Input: {
    Headers?: t.ExactType<t.Type<H, unknown>>;
    Params?: t.ExactType<t.Type<P, unknown>>;
    Query?: t.Type<Q, unknown>;
    Body?: t.Type<B, unknown>;
  };
  Output: t.Type<R, unknown> | t.ArrayType<t.Type<R>>;
}

/**
 * Constructor function for an endpoint
 *
 * @param Method The [HTTPMethod] used
 * @param Path The absolute path the enpoint can be reached at
 * @param Input A dictionary of `io-ts` codecs representing the type of the expected input
 * @param Output An `io-ts` codec representing the type of the expected output
 */
export function Endpoint<O extends EndpointOptions, P, H, Q, B, R>(
  e: Endpoint<O, P, H, Q, B, R>
): Endpoint<O, P, H, Q, B, R> {
  return {
    Opts: e.Opts,
    Method: e.Method,
    Path: e.Path,
    Input: e.Input,
    Output: e.Output,
  };
}
