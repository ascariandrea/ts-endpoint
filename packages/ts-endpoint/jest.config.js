module.exports = {
  verbose: true,
  preset: 'ts-jest',
  globals: {
    'ts-jest': {
      tsConfig: './tsconfig.json',
    },
  },
  testEnvironment: 'node',
  roots: ['<rootDir>/src'],
  testMatch: ['**/?(*.)(spec).ts?(x)'],
};
