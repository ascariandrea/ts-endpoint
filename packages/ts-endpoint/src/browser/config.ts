import * as t from 'io-ts';
import { PositiveNumber } from '../shared/PositiveNumber';

export const HTTPProtocol = t.keyof(
  {
    http: null,
    https: null,
  },
  'HTTPProtocol'
);
export type HTTPProtocol = t.TypeOf<typeof HTTPProtocol>;

export const HTTPClientConfig = t.strict(
  {
    protocol: HTTPProtocol,
    host: t.string,
    port: t.union([PositiveNumber, t.undefined]),
  },
  'HTTPClientConfig'
);
export type HTTPClientConfig = t.TypeOf<typeof HTTPClientConfig>;
export type StaticHTTPClientConfig = t.OutputOf<typeof HTTPClientConfig>;
