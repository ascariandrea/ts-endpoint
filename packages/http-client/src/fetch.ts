import {
  type Codec,
  type DecodeCodecFn,
  type EndpointErrors,
  IOError,
  type MinimalEndpointInstance,
  NetworkErrorStatus,
} from '@ts-endpoint/core';
import { toArray } from 'fp-ts/Record';
import { findFirst } from 'fp-ts/lib/Array.js';
import * as O from 'fp-ts/lib/Option.js';
import * as TA from 'fp-ts/lib/TaskEither.js';
import { pipe } from 'fp-ts/lib/function.js';
import qs from 'qs';
import { type FetchClient, GetHTTPClient, type GetHTTPClientOptions } from './GetHTTPClient.js';
import { type HTTPClientConfig } from './config.js';

declare module './HKT.js' {
  interface URItoKind<A> {
    IOError: A extends Record<string, Codec<any, any, never>> ? IOError<A> : never;
  }
}

export const GetFetchHTTPClient = <A extends { [key: string]: MinimalEndpointInstance }>(
  config: HTTPClientConfig,
  endpoints: A,
  options: GetHTTPClientOptions
) => GetHTTPClient(config, endpoints, useBrowserFetch, options);

const getResponseJson = (r: Response, ignoreNonJSONResponse: boolean) => {
  if (r.body === null) {
    TA.left(undefined);
  }
  const jsonResponse = TA.tryCatch(
    () => r.json(),
    () =>
      new IOError(r.statusText, {
        status: r.status.toString(),
        kind: 'ServerError',
        meta: { message: 'response is not a json.' },
      })
  );

  return ignoreNonJSONResponse
    ? pipe(
        jsonResponse,
        TA.alt(() => TA.of(undefined))
      )
    : jsonResponse;
};

export const useBrowserFetch = <E extends MinimalEndpointInstance>(
  baseURL: string,
  e: E,
  options: GetHTTPClientOptions
): FetchClient<E, 'IOError'> => {
  return (i: any) => {
    const path = `${baseURL}${e.getPath(i?.Params)}${i?.Query ? `?${qs.stringify(i?.Query)}` : ''}`;

    const body = i?.Body;
    const headers = { ...(i?.Headers ?? {}), ...options?.defaultHeaders };

    const response = pipe(
      TA.tryCatch(
        () =>
          fetch(path, {
            ...{
              headers,
              method: e.Method,
            },
            ...(i?.Body !== undefined ? { body: JSON.stringify(body) } : {}),
          }),
        () => new IOError('Network Error', { kind: 'NetworkError', status: NetworkErrorStatus })
      ),
      TA.chain((r: Response) => {
        const responseJson: TA.TaskEither<any, any> = TA.tryCatch(
          () => (r.body === null ? Promise.resolve(undefined) : r.json()),
          () => {
            return new IOError(r.statusText, {
              status: r.status.toString(),
              kind: 'ServerError',
              meta: { message: 'response is not a json.' },
            });
          }
        );

        const responseJsonWithDefault = options?.ignoreNonJSONResponse
          ? pipe(
              responseJson,
              TA.alt(() => {
                return TA.of(undefined);
              })
            )
          : responseJson;

        if (!r.ok) {
          if (e.Errors !== undefined) {
            const actualKnownError = pipe(
              e.Errors as EndpointErrors<string, Codec<any, any>>,
              toArray,
              findFirst(([status]) => {
                return status === r.status.toString();
              })
            );

            if (O.isSome(actualKnownError)) {
              const decodeAsDecodeRecord = options.decode as DecodeCodecFn<any>;
              const parsedKnownError = pipe(
                responseJsonWithDefault,
                TA.map((body) => decodeAsDecodeRecord(actualKnownError.value[1])(body)),
                TA.chain((validation) => {
                  return pipe(
                    TA.fromEither(validation),
                    TA.fold(
                      (error) => {
                        return TA.left(error);
                      },

                      (parsedError) => {
                        return TA.left(
                          new IOError<any>(r.statusText, {
                            kind: 'KnownError',
                            status: r.status.toString(),
                            body: parsedError,
                          })
                        );
                      }
                    )
                  );
                })
              );

              return parsedKnownError;
            }
          }

          if (r.status >= 400 && r.status <= 451) {
            return pipe(
              getResponseJson(r, options?.ignoreNonJSONResponse ?? false),
              TA.chain((meta) => {
                return TA.left(
                  new IOError(r.statusText, {
                    kind: 'ClientError',
                    meta,
                    status: r.status.toString(),
                  })
                );
              })
            );
          } else {
            return pipe(
              getResponseJson(r, options?.ignoreNonJSONResponse ?? false),
              TA.chain((meta) => {
                return TA.left(
                  new IOError(r.statusText, {
                    kind: 'ServerError',
                    meta,
                    status: r.status.toString(),
                  })
                );
              })
            );
          }
        }

        const decodeAsOutputDecoder = options.decode as DecodeCodecFn<any>;

        const res = pipe(
          responseJsonWithDefault,
          TA.chain((json) => {
            return pipe(
              options?.mapInput ? options.mapInput(json) : json,
              decodeAsOutputDecoder(e.Output),
              TA.fromEither
            );
          })
        );

        return res;
      }),
      TA.mapLeft((err) => {
        if (options?.handleError !== undefined) {
          return options.handleError(err, e);
        }
        return err;
      })
    );

    return response;
  };
};
