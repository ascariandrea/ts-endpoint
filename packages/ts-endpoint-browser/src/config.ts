import * as t from 'io-ts';
import { PositiveNumber } from '../../ts-shared/src/PositiveNumber';

export const HTTPProtocol = t.keyof(
  {
    http: null,
    https: null,
  },
  'HTTPProtocol'
);
export type HTTPProtocol = t.TypeOf<typeof HTTPProtocol>;

export const HTTPClientConfig = t.exact(
  t.intersection([
    t.type({
      protocol: HTTPProtocol,
      host: t.string,
    }),
    t.partial({ port: t.union([PositiveNumber, t.undefined]) }),
  ]),
  'HTTPClientConfig'
);
export type HTTPClientConfig = t.TypeOf<typeof HTTPClientConfig>;
export type StaticHTTPClientConfig = t.OutputOf<typeof HTTPClientConfig>;
