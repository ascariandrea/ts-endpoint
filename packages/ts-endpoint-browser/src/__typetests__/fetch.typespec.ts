import { GetFetchHTTPClient } from '../fetch';
import { Endpoint } from 'ts-endpoint';
import { StaticHTTPClientConfig } from '../config';
import * as t from 'io-ts';
import { InferFetchResult } from '..';
import { right } from 'fp-ts/lib/Either';

const options: StaticHTTPClientConfig = {
  protocol: 'http',
  host: 'http://test',
  port: 2020,
};

const endpoints = {
  prova: Endpoint({
    Input: {
      Query: { color: t.string },
      Params: { id: t.string },
    },
    Method: 'GET',
    getPath: ({ id }) => `users/${id}/crayons`,
    Output: t.type({ crayons: t.array(t.string) }),
    Opts: { stringifyBody: true },
  }),
};

const fetchClient = GetFetchHTTPClient(options, endpoints, {
  handleError: (err, e) => err
});

// @dts-jest:fail:snap should not allow empty calls
fetchClient.prova();

// @dts-jest:fail:snap should not allow empty-object as input
fetchClient.prova({});

// @dts-jest:fail:snap should not allow calls not specifing Query (if needed)
fetchClient.prova({
  Params: { id: '123' },
});

fetchClient.prova({
  // @dts-jest:fail:snap should not allow calls specifing Params not declared in the endpoint
  Params: { id: '123', foo: 'baz' },
  Query: { color: 'marrone' },
});

fetchClient.prova({
  Params: { id: '123' },
  Query: { color: 'marrone' },
  // @dts-jest:fail:snap should not allow to add Body when not declared in the endpoint
  Body: { foo: 'baz' },
});

// @dts-jest:fail:snap InferFetchResult should return the type of the Endpoint.Output
const provaResultWrong: InferFetchResult<typeof fetchClient.prova> = right({});

// @dts-jest:pass:snap InferFetchResult should return the type of the Endpoint.Output
const provaResultCorrect: InferFetchResult<typeof fetchClient.prova> = right({
  crayons: ['brown'],
});
