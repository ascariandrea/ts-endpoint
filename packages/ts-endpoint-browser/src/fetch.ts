import {
  EndpointInstance,
  TypeOfEndpointInstance,
  Endpoint,
  HTTPMethod,
} from 'ts-endpoint/src/Endpoint';
import { HTTPClientConfig, StaticHTTPClientConfig } from './config';
import { pipe } from 'fp-ts/lib/pipeable';
import { PathReporter } from 'io-ts/lib/PathReporter';
import * as TA from 'fp-ts/lib/TaskEither';
import * as t from 'io-ts';
import qs from 'qs';
import { IOError, DecodeErrorStatus, NetworkErrorStatus } from 'ts-shared/lib/errors';
import { HTTPClient, GetHTTPClient, FetchClient } from '.';
import { left } from 'fp-ts/lib/Either';

export const GetFetchHTTPClient = <A extends { [key: string]: EndpointInstance<any> }>(
  config: HTTPClientConfig | StaticHTTPClientConfig,
  endpoints: A,
  defaultHeaders?: { [key: string]: string }
): HTTPClient<A, IOError> => GetHTTPClient(config, endpoints, useBrowserFetch, defaultHeaders);

const getResponseJson = (r: Response) => {
  return TA.tryCatch(
    () => r.json(),
    () =>
      new IOError(r.status, r.statusText, {
        kind: 'ServerError',
        meta: { message: 'response is not a json.' },
      })
  );
};

export const useBrowserFetch = <
  M extends HTTPMethod,
  O extends t.Type<any, any, any>,
  H extends { [k: string]: t.Type<any, any, any> } | undefined = undefined,
  Q extends { [k: string]: t.Type<any, any, any> } | undefined = undefined,
  B extends t.Type<any, any, any> | undefined = undefined,
  P extends { [k: string]: t.Type<any, any, any> } | undefined = undefined
>(
  baseURL: string,
  e: EndpointInstance<Endpoint<M, O, H, Q, B, P>>,
  defaultHeaders?: { [key: string]: string }
): FetchClient<EndpointInstance<Endpoint<M, O, H, Q, B, P>>, IOError> => {
  return (i: TypeOfEndpointInstance<EndpointInstance<Endpoint<M, O, H, Q, B, P>>>['Input']) => {
    const path = `${baseURL}${e.getPath(i?.Params ?? {})}${
      i.Query ? `?${qs.stringify(i.Query)}` : ''
    }`;
    const body = e.Opts?.stringifyBody ? qs.stringify(i.Body) : i.Body;
    const headers = { ...i.Headers, ...defaultHeaders };

    const response = pipe(
      TA.tryCatch(
        () =>
          fetch(path, {
            ...{
              headers,
              method: e.Method,
            },
            ...(i.Body !== undefined ? { body: JSON.stringify(body) } : {}),
          }),
        () => new IOError(NetworkErrorStatus, 'Network Error', { kind: 'NetworkError' })
      ),
      TA.chain((r: Response) => {
        const responseJson: TA.TaskEither<any, any> = TA.tryCatch(
          () => r.json(),
          () =>
            new IOError(r.status, r.statusText, {
              kind: 'ServerError',
              meta: { message: 'response is not a json.' },
            })
        );

        if (!r.ok) {
          if (r.status >= 400 && r.status <= 451) {
            return pipe(
              getResponseJson(r),
              TA.chain((meta) =>
                TA.left(new IOError(r.status, r.statusText, { kind: 'ClientError', meta }))
              )
            );
          } else {
            return pipe(
              getResponseJson(r),
              TA.chain((meta) =>
                TA.left(new IOError(r.status, r.statusText, { kind: 'ServerError', meta }))
              )
            );
          }
        }

        e.Output;

        const res = pipe(
          responseJson,
          TA.chain((json) =>
            pipe(
              TA.fromEither(e.Output.decode(json)),
              TA.mapLeft(
                (errors) =>
                  new IOError(
                    DecodeErrorStatus,
                    `Error decoding server response: ${PathReporter.report(left(errors))}`,
                    {
                      kind: 'DecodingError',
                      errors,
                    }
                  )
              )
            )
          )
        );

        return res;
      })
    );

    return response;
  };
};
