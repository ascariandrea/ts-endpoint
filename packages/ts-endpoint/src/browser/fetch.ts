import { Endpoint } from '../Endpoint';
import { HTTPClientConfig, StaticHTTPClientConfig } from './config';
import { pipe } from 'fp-ts/lib/pipeable';
import { PathReporter } from 'io-ts/lib/PathReporter';
import * as TA from 'fp-ts/lib/TaskEither';
import * as t from 'io-ts';
import qs from 'qs';
import { IOError, DecodeErrorStatus, NetworkErrorStatus } from './errors';
import { HTTPClient, GetHTTPClient, FetchClient, FetchInput } from '.';
import { Either, left } from 'fp-ts/lib/Either';

export const GetFetchHTTPClient = <
  A extends { [key: string]: Endpoint<any, any, any, any, any, any> }
>(
  config: HTTPClientConfig | StaticHTTPClientConfig,
  endpoints: A,
  defaultHeaders?: { [key: string]: string }
): HTTPClient<A, IOError> => GetHTTPClient(config, endpoints, useBrowserFetch, defaultHeaders);

const useBrowserFetch = <E extends Endpoint<any, any, any, any, any, any>>(
  baseURL: string,
  e: E,
  defaultHeaders?: { [key: string]: string }
): FetchClient<E, IOError> => {
  return (i: FetchInput<E>) => {
    // TODO: try to get rid of this
    const anyArgs: any = i;

    const path = `${baseURL}/${e.getPath(anyArgs?.Params ?? {})}${
      anyArgs.Query ? `?${qs.stringify(anyArgs.Query)}` : ''
    }`;
    const body = e.Opts?.stringifyBody ? qs.stringify(anyArgs.Body) : anyArgs.Body;
    const headers = { ...anyArgs.Headers, ...defaultHeaders };

    const response = pipe(
      TA.tryCatch(
        () =>
          fetch(path, {
            ...{
              headers,
              method: e.Method,
            },
            ...(anyArgs.Body !== undefined ? { body } : {}),
          }),
        () => new IOError(NetworkErrorStatus, 'Network Error', { kind: 'NetworkError' })
      ),
      TA.chain((r: Response) => {
        if (!r.ok) {
          if (r.status >= 400 && r.status <= 451) {
            return TA.left<IOError, t.TypeOf<E['Output']>>(
              new IOError(r.status, r.statusText, { kind: 'ClientError' })
            );
          }
          return TA.left<IOError, t.TypeOf<E['Output']>>(
            new IOError(r.status, r.statusText, { kind: 'ServerError' })
          );
        }

        const res = pipe(
          TA.taskEither.fromTask<t.Errors, any>(() => r.json()),
          TA.alt(() => TA.right({})),
          TA.chain((json) =>
            TA.fromEither(e.Output.decode(json) as Either<t.Errors, t.TypeOf<E['Output']>>)
          ),
          TA.mapLeft(
            (errors) =>
              new IOError(
                DecodeErrorStatus,
                `Error decoding server response: ${PathReporter.report(left(errors))}`,
                {
                  kind: 'DecodingError',
                  errors,
                }
              )
          )
        );

        return res;
      })
    );

    return response;
  };
};
