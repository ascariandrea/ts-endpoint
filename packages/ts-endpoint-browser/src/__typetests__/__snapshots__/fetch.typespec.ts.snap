// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InferFetchResult should return the type of the Endpoint.Output (type) should match snapshot 1`] = `
"Type 'Either<EndpointInstanceError<EndpointInstance<Endpoint<\\"GET\\", TypeC<{ crayons: ArrayC<StringC>; }>, undefined, { color: StringC; }, undefined, { id: StringC; }, undefined>>>, {}>' is not assignable to type 'Either<EndpointInstanceError<EndpointInstance<Endpoint<\\"GET\\", TypeC<{ crayons: ArrayC<StringC>; }>, undefined, { color: StringC; }, undefined, { id: StringC; }, undefined>>>, { ...; }>'.
  Type 'Right<{}>' is not assignable to type 'Either<EndpointInstanceError<EndpointInstance<Endpoint<\\"GET\\", TypeC<{ crayons: ArrayC<StringC>; }>, undefined, { color: StringC; }, undefined, { id: StringC; }, undefined>>>, { ...; }>'.
    Type 'Right<{}>' is not assignable to type 'Right<{ crayons: string[]; }>'.
      Property 'crayons' is missing in type '{}' but required in type '{ crayons: string[]; }'."
`;

exports[`InferFetchResult should return the type of the Endpoint.Output (type) should match snapshot 2`] = `"any"`;

exports[`should not allow calls not specifing Query (if needed) (type) should match snapshot 1`] = `
"Argument of type '{ Params: { id: string; }; }' is not assignable to parameter of type '{ Params: { id: string; }; Headers?: never; Query: { color: string; }; Body?: never; }'.
  Property 'Query' is missing in type '{ Params: { id: string; }; }' but required in type '{ Params: { id: string; }; Headers?: never; Query: { color: string; }; Body?: never; }'."
`;

exports[`should not allow calls specifing Params not declared in the endpoint (type) should match snapshot 1`] = `
"Type '{ id: string; foo: string; }' is not assignable to type '{ id: string; }'.
  Object literal may only specify known properties, and 'foo' does not exist in type '{ id: string; }'."
`;

exports[`should not allow empty calls (type) should match snapshot 1`] = `"Expected 1 arguments, but got 0."`;

exports[`should not allow empty-object as input (type) should match snapshot 1`] = `
"Argument of type '{}' is not assignable to parameter of type '{ Params: { id: string; }; Headers?: never; Query: { color: string; }; Body?: never; }'.
  Type '{}' is missing the following properties from type '{ Params: { id: string; }; Headers?: never; Query: { color: string; }; Body?: never; }': Params, Query"
`;

exports[`should not allow to add Body when not declared in the endpoint (type) should match snapshot 1`] = `"Type 'string' is not assignable to type 'never'."`;

exports[`you can access KnownErrors with the correct typeguard (type) should match snapshot 1`] = `"string"`;

exports[`you cannot access KnownErrors without the correct typeguard (type) should match snapshot 1`] = `"Property 'baz' does not exist on type '{ foo: string; }'."`;
